packages/core/admin/admin/src/pages/MarketplacePage/components/NpmPackagesGrid/index.js

```diff
@@ -1,29 +1,60 @@
-import React, { useCallback } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
+import { useIntl } from 'react-intl';
+import { AnErrorOccurred } from '@strapi/helper-plugin';
 import { Grid, GridItem } from '@strapi/design-system/Grid';
+import { Flex } from '@strapi/design-system/Flex';
+import { Loader } from '@strapi/design-system/Loader';
 import NpmPackageCard from '../NpmPackageCard';
+import EmptyNpmPackageSearch from '../EmptyNpmPackageSearch';
 
 const NpmPackagesGrid = ({
+  status,
   npmPackages,
   installedPackageNames,
   useYarn,
   isInDevelopmentMode,
   npmPackageType,
   strapiAppVersion,
+  debouncedSearch,
 }) => {
-  // Check if an individual package is in the dependencies
-  const isAlreadyInstalled = useCallback(
-    (npmPackageName) => installedPackageNames.includes(npmPackageName),
-    [installedPackageNames]
+  const { formatMessage } = useIntl();
+
+  if (status === 'error') {
+    return (
+      <Flex paddingTop={8}>
+        <AnErrorOccurred />
+      </Flex>
+    );
+  }
+
+  if (status === 'loading') {
+    return (
+      <Flex justifyContent="center" paddingTop={8}>
+        <Loader>Loading content...</Loader>
+      </Flex>
+    );
+  }
```

Comment:
It's true that redwood solution is interesting

Having a bunch of loaders everywhere was also one of the reasons for React suspense, but to use it with data fetching we'll have to wait for React 18 in Strapi, so v5 unfortunately

Reply:
oh I was thinking something like this, but not sure if it makes sense, and it could be another PR

```
// definition
const { status, data } = props.stuffFromQuery

// conditions for handling errors, empty, and loading

return props.children

// some parent compo
const stuffFromQuery = useQuery()
<FetchResult useQuery={stuffFromUseQuery}>
// bunch of jsx for success
</FetchResult>
```

Reply:
I really like the idea 👍 

I assume point would be to use it throughout the application, and for that we'd probably need to prepare a topic at the frontend sync. What do you think?

Reply:
Yup sounds like a plan!